File: 2022/July/22 July.py
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if(left == right) return head;
        ListNode newHead = head, newTail = head;
        for(int i = 1; i < left; i++) {
            newTail = head;
            head = head.next;
        }
        ListNode revHead = head, revTail = head;
        for(int i = left; i < right+1; i++) {
            ListNode temp = head.next;
            head.next = revHead;
            revHead = head;
            head = temp;
        }
        newTail.next = revHead;
        revTail.next = head;
        if(left == 1) return revHead;
        return newHead;
    }
}

File: 2022/July/22 July.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode insertLast(ListNode tail, ListNode node) {
        if(tail == null) {
            node.next = null;
            return node;
        }
        tail.next = node;
        node.next = null;
        return node;
    }
    public ListNode partition(ListNode head, int x) {
        ListNode less = null, lessTail = null, more = null, moreTail = null;
        while(head != null) {
            ListNode temp = head.next;
            if(head.val < x) {
                lessTail = insertLast(lessTail, head);
                if(less == null) less = lessTail;
            } else {
                moreTail = insertLast(moreTail, head);
                if(more == null) more = moreTail;
            }
            head = temp;
        }
        if(less != null) {
            head = less;
            lessTail.next = more;
        } else {
            head = more;
        }
        return head;
    }
}
